datafusion/expr/src/signature.rs

```diff
@@ -346,13 +346,60 @@ impl Signature {
     }
 }
 
-/// Monotonicity of the `ScalarFunctionExpr` with respect to its arguments.
-/// Each element of this vector corresponds to an argument and indicates whether
-/// the function's behavior is monotonic, or non-monotonic/unknown for that argument, namely:
-/// - `None` signifies unknown monotonicity or non-monotonicity.
-/// - `Some(true)` indicates that the function is monotonically increasing w.r.t. the argument in question.
-/// - Some(false) indicates that the function is monotonically decreasing w.r.t. the argument in question.
-pub type FuncMonotonicity = Vec<Option<bool>>;
+/// Monotonicity of a function with respect to its arguments.
+///
+/// A function is [monotonic] if it preserves the relative order of its inputs.
+///
+/// [monotonic]: https://en.wikipedia.org/wiki/Monotonic_function
+#[derive(Debug, Clone)]
+pub enum FuncMonotonicity {
+    /// not monotonic or unknown monotonicity
+    None,
+    /// Increasing with respect to all of its arguments
+    Increasing,
+    /// Decreasing with respect to all of its arguments
+    Decreasing,
+    /// Each element of this vector corresponds to an argument and indicates whether
+    /// the function's behavior is monotonic, or non-monotonic/unknown for that argument, namely:
+    /// - `None` signifies unknown monotonicity or non-monotonicity.
+    /// - `Some(true)` indicates that the function is monotonically increasing w.r.t. the argument in question.
+    /// - Some(false) indicates that the function is monotonically decreasing w.r.t. the argument in question.
+    Mixed(Vec<Option<bool>>),
+}
+
+impl PartialEq for FuncMonotonicity {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (FuncMonotonicity::None, FuncMonotonicity::None) => true,
+            (FuncMonotonicity::Increasing, FuncMonotonicity::Increasing) => true,
+            (FuncMonotonicity::Decreasing, FuncMonotonicity::Decreasing) => true,
+            (FuncMonotonicity::Mixed(vec1), FuncMonotonicity::Mixed(vec2)) => {
+                vec1 == vec2
+            }
+            _ => false,
+        }
+    }
+}
+
+impl FuncMonotonicity {
+    pub fn matches(&self, other: &Self) -> bool {
```

Comment:
Note: The `matches` method isn't currently being used anywhere in the codebase. It has been added to make comparisons easier for any future purposes.

---
Also, I don't think we need the arity of the function anymore for comparison b/w `FuncMonotonicity::Mixed` and other types.

For eg: `FuncMonotonicity::Increasing` is "Increasing with respect to all of its arguments" which is the same as the `FuncMonotonicity::Mixed`'s inner vector being `Some(true)` for all elements from what I understand.